mono/metadata/w32file-win32.c

```diff
@@ -81,23 +81,63 @@ mono_w32file_delete (const gunichar2 *name)
 	return res;
 }
 
+static void
+abort_w32_synchronous_io (HANDLE hThread)
+{
+	CancelSynchronousIo (hThread);
+}
+
 gboolean
-mono_w32file_read(gpointer handle, gpointer buffer, guint32 numbytes, guint32 *bytesread)
+mono_w32file_read (gpointer handle, gpointer buffer, guint32 numbytes, guint32 *bytesread)
 {
 	gboolean res;
+	gboolean interrupted;
+	HANDLE duplicate;
+
+	duplicate = OpenThread (THREAD_ALL_ACCESS, FALSE, GetCurrentThreadId ());
```

Comment:
MSDN says  THREAD_TERMINATE, not THREAD_ALL_ACCESS.
There is also MAXIMUM_ALLOWED if  you want the maximum you can get and  be prepared for access denied on later use.

Reply:
Actually  since you are using  GetCurrentThreadId anyway, just use GetCurrentThread(). It is very much faster (just returns a constant -2) and grants all access. No need to duplicate or close, as long as you  are only using it on the current thread. If are you passing it across thread/process boundaries, then use DuplicateHandle(GetCurrentThread()), not OpenThread.

Reply:
Hmm, I initially had `GetCurrentThread` and the call to `CancelSynchronousIo` was failing. Let me revisit.

Reply:
I can understand that -- the Cancel happens from another thread, right? In that case, DuplicateHandle(GetCurrentThread()) is slightly preferred. Same number of syscalls either way -- open(GetCurrentThreadId()) vs. duplicate(GetCurrentThread()). But Duplicate is kinda better. Granted it takes more parameters. GetCurrentThreadId and GetCurrentThread are both fast/small, not syscalls. GetCurrentThreadId just reads memory,  GetCurrentThread just returns -2.